Enable multiple instances of MGT to exist

shweaver-MSFT opened the pull request:
<!-- Review contributing guidelines before creating PRs -->
<!-- https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/main/CONTRIBUTING.md -->

Closes #229 <!-- REQUIRED: Add the issue number (ex: #12) so the issue is automatically closed when PR is merged -->

### PR Type
<!-- Please uncomment one ore more that apply to this PR -->

- Bugfix
<!-- - Feature -->
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Other... Please describe: -->

### Description of the changes
The customElements registry does not allow for duplicate class registrations, which causes problems in multi-instance scenarios such as SharePoint. 

~To fix this, I've added a new class, `utils/ComponentRegistry`, to handle the registration logic for all components. To apply the new logic path, I stripped out all of the `@customElement('mgt-foo');` and replaced them with my new decorator `@registeredComponent('mgt-foo');`.~

~I added the scoped elements package (@open-wc/scoped-elements) and applied it to the BaseComponent. The effect is that all internally used elements are protected from naming collisions and duplicate definitions.~

UPDATE

To fix this, I've added a new class, `utils/ComponentRegistry`, to handle the registration logic for all components. To apply the new logic path, I stripped out all of the `@customElement('mgt-foo');` and instead added the component registration declarations to the `index.ts` for each component.

I also added the scoped elements package (@open-wc/scoped-elements) to surface scoped versions of the components to developers. This enables the developer to use scoped instances of the components which should prevent accidental component naming/version collisions at runtime in environments like SharePoint.

### PR checklist
- [x] Project builds (`npm run build`) and changes have been tested in supported browsers (including IE11)
- [x] All public classes and methods have been documented
- [x] Added appropriate [documentation](https://github.com/microsoftgraph/microsoft-graph-docs/tree/master/concepts/toolkit) [target branch `mgt/next` for new features]. Docs PR: https://github.com/microsoftgraph/microsoft-graph-docs/pull/9209
- [x] License header has been added to all new source files (`npm run setLicense`)
- [x] Contains **NO** breaking changes

### Other information
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->

shweaver-MSFT commented:
@nmetulev, I love your suggestions! Sounds like a solid strategy and the opt-in pattern for scoped elements is smart. My only nitpick is putting an index.ts in every component folder. Mostly just because I'm lazy, and that's a lot of files! :P

Could we simply do the registrations in the components.ts? We already reference each individual component in that file, so it might be a good *singular* place to put it. It would also be more inline with the scoped component declarations.

Also, the scoped folder with index.html could also probably be replaced with single file, scoped.ts. Same effect, but cleaner imo.

I'll try it out and update the PR :)

nmetulev commented:
I agree it would be tedious to add the index files to all of them, but there is an awesome end user benefit to consider. The index file does not have to be specified when doing an import. For example, to register a single component, today a developer would need to do:

```ts
import "@microsoft/mgt/dist/es6/components/mgt-person/mgt-person.js"
```

with a per component index.ts, this would turn into:
```ts
import "@microsoft/mgt/dist/es6/components/mgt-person"
```

scoped would be similar
```ts
import "@microsoft/mgt/dist/es6/components/mgt-person/scoped"
```

The per component index.ts file is an important optimization IMO as it allows the developer to import just the components they needs and for bundlers to tree shake the rest.

I wish I could take credit for this, but it seems this is a standard practice for building components.

shweaver-MSFT commented:
Something about the way the index files were changed has broken the way the web component analyzer detects the components. This prevents the custom-elements.json from being generated thus breaking storybook. Need to figure out what's going on

shweaver-MSFT commented:
Turns out the web component analyzer is expecting the `@customElement(...)` or `customElements.define(...)` to determine where the components are. By moving the component declarations to the index.ts files, wca can no longer detect *any* of the components.

No worries though. Just need to add `@element` tags to the docs for each component. Wca seems to work fine with that.

shweaver-MSFT commented:
Ok, looks like the PR is in good shape. Everything discussed so far has been fixed up. 👍

nmetulev commented on packages/mgt/src/mock/mgt-mock-provider.ts:
```diff
@@ -5,18 +5,18 @@
  * @class MgtMockProvider
  * @extends {LitElement}
  */
-@customElement('mgt-mock-provider')
```
This removes the mock-provider registration without adding an index.ts file. There is no way to use it unless the dev registers it manually.

IMO, mgt-mock-provider should remain as is as it's only used for development

shweaver-MSFT replied:
Oops, looks like I missed that one. But I agree with you, it doesn't need an extra scoped definition.

shweaver-MSFT commented on packages/mgt/src/components/baseComponent.ts:
```diff
@@ -5,8 +5,10 @@
+import { ScopedElementsMixin } from '@open-wc/scoped-elements';
+import { LitElement, PropertyValues } from 'lit-element';
 import { Providers } from '../Providers';
+import { ComponentRegistry } from '../utils/ComponentRegistry';
```
Nope

shweaver-MSFT commented on packages/mgt/src/components/mgt-login/scoped/index.ts:
```diff
@@ -0,0 +1,15 @@
+import { ScopedElementsMixin } from '@open-wc/scoped-elements';
+import { ComponentRegistry } from '../../../utils/ComponentRegistry';
+import { MgtLogin } from '../mgt-login';
```
No it doesn't. mgt-login is the file name being referenced in the previous folder. I just tested to make sure and the index.ts isn't executed. I'd have to do `'../index'` for that effect.

nmetulev replied:
You are right, I read it wrong. Thanks for clarifying.

shweaver-MSFT commented:
@nmetulev 

>1. npm run storybook seems to be broken. Adding the @webcomponents/webcomponentsjs to root package.json seems to solve the issue

I noticed this too, good catch. I fixed it locally and didn't commit after testing :/

> 2. There are some css rules that remain referencing the components directly, will these cause any issues?

Accidental leftovers. I removed them.

About this part:
> It's not clear how scoping would work as the tags will remain the same for scoped components.

I agree, it's not entirely clear in my mind yet either. When I think about the scenario, the instances of MGT are running in isolated iframes (at least in SharePoint), which prevents js from colliding which the exception of the global element registry. Adding the loader twice in the same frame is a bit odd and probably shouldn't work imo. 

I'm thinking theoretically, but if running in SharePoint then none of the original tag names can be trusted, even inside the components. I'll re-add the prefix logic. In addition though, it might make sense for components to use other scoped components internally. This would ensure that a V1 defined component doesn't get consumed in a different V-whatever component because the V1 happened to be registered first. I worry that adding prefixes or whatever to the tag names won't flow into the templates which just use pure `mgt-` tag names names. I know that definitely breaks the "opt-in" approach, but it makes more sense to me based on the scenario. I can't think of any other way to do it.

Any thoughts/ideas?

shweaver-MSFT commented:
## We need your help

This is a tough one. After much discussion on the team, we aren't sure exactly the correct approach to solve this issue. 

This attempt is built on the assumption that all MGT instances would be used in an iframe, but to my chagrin this isn't always the case. In addition, support for scoped element registries is not yet adopted by the W3C, so I had to work with a [polyfill](https://open-wc.org/scoped-elements/). Some good discussion on the topic can be found [here](https://github.com/w3c/webcomponents/issues/716).

If there are any particularly savvy developers out there, we welcome your expertise 😊

frank-dspeed commented:
@shweaver-MSFT i can help with that there are other component patterns you do not even need to use document.define I am Senior Engineer we can talk about the right method for your case if you like you can email me or schedule a meeting via hangouts or you choose the platform to talk about that.

I am more then happy to help with that as the custom element registry pattern is bad and i do not want that it spreads around.

frank-dspeed commented:
@shweaver-MSFT https://dev.to/frankdspeed/the-html-component-pattern-hcp-3pkp little demo of the topic showing how to init a custom Element even without registry

frank-dspeed commented:
## Some Basics that help you

There is a global ESM Registry it is sorted by path so it is uniq to the file name if you import something more then once you always get the same Module back the ESM Cache is on per window base it is the same registrie that the serviceWorkers and other Th models do have access to.

so when we import a customElement definition (class) 

we can apply it manual via sync or defered script tags as they have access to currentScript which points to the dom node the script is running in this is scoped we can even do a clean up after we upgraded the element we can delet our script from the dom 
it is ultra fast no registrie envolved when we render some new components we can also attach the script tags again to init them or we can init them in the same step when we apply the definition.

frank-dspeed commented:
# Next Steps
Remove @customElement('mgt-get') and all other calls to customElement Register make this a lib of element definitions that a user can register under his own tag.

shweaver-MSFT commented:
Hi @frank-dspeed, thanks for your feedback! I admit, I have had some trouble finding the right pattern for integrating scoped custom elements. As a team, we decided to hold off on this so we can focus on other priorities for our next release. 

Would you be willing to get involved and contribute a PR with the integration? Your expertise in this area would be greatly appreciated. It will otherwise take some time before I have a chance to re-investigate and implement.

We also have some concerns around performance. Do you have any insight into the potential performance impact of adopting scoped custom elements across the entire project?

frank-dspeed commented:
@shweaver-MSFT can you supply a example rep that shows the need for a own registrie at present i did not spot the cases i found only your custom-elements and if you do not register them inside the elements you will never need one.
